### PR TITLE
Fixes for transducer code

### PR DESCRIPTION
**Pull request by lorenlugosch**
(Can also fix TIMIT numbers)

### CODE REVIEW

**Comment by aheba** [comment]
Done, LGTM

**Comment by aheba** [comment]
Loren, I can quickly run wav2vec with Transducer...
let me a day to do that before merging...

**Comment by lorenlugosch** [comment]
Looks like there's an issue where `TransducerLoss` works but `transducer_loss` doesn't, Abdel's looking into it

**Comment by lorenlugosch** [comment]
@aheba in the `TransducerBeamSearch` example (https://github.com/speechbrain/speechbrain/blob/develop/speechbrain/decoders/transducer.py) you use a bidirectional GRU for the predictor? That can't be right, unless I'm misunderstanding this code?

**Comment by aheba** [comment]
hey @lorenlugosch , let me share with you the wav2vec transducer example..

**Comment by lorenlugosch** [comment]
Do you have PER numbers for the wav2vec2 TIMIT model to update the README?

**Comment by aheba** [comment]
I'm on it, let me share in few hours the PER:
```
epoch: 1, lr_adam: 3.00e-04, lr_wav2vec: 1.00e-04 - train loss: 7.59e-01 - valid loss: 2.05e-01, valid PER: 52.55
epoch: 2, lr_adam: 3.00e-04, lr_wav2vec: 1.00e-04 - train loss: 2.23e-01 - valid loss: 9.06e-02, valid PER: 23.94
epoch: 3, lr_adam: 3.00e-04, lr_wav2vec: 1.00e-04 - train loss: 1.47e-01 - valid loss: 6.70e-02, valid PER: 12.70
epoch: 4, lr_adam: 3.00e-04, lr_wav2vec: 1.00e-04 - train loss: 1.19e-01 - valid loss: 6.16e-02, valid PER: 10.57
epoch: 5, lr_adam: 3.00e-04, lr_wav2vec: 1.00e-04 - train loss: 1.05e-01 - valid loss: 5.84e-02, valid PER: 9.55
epoch: 6, lr_adam: 3.00e-04, lr_wav2vec: 1.00e-04 - train loss: 9.52e-02 - valid loss: 5.90e-02, valid PER: 8.89
epoch: 7, lr_adam: 3.00e-04, lr_wav2vec: 1.00e-04 - train loss: 8.78e-02 - valid loss: 5.81e-02, valid PER: 8.24
epoch: 8, lr_adam: 3.00e-04, lr_wav2vec: 1.00e-04 - train loss: 7.97e-02 - valid loss: 5.96e-02, valid PER: 7.97
```

**Comment by lorenlugosch** [comment]
!!

**Comment by aheba** [comment]
> !!

The model is still on training phase.. !

**Comment by lorenlugosch** [comment]
Oh yeah, I thought those were test PERs for a moment. Still pretty cool :)

**Comment by aheba** [comment]
hey @lorenlugosch , due to the small GPU I have (1GPU V100 16GB).. my RTX8000 server is still down.. So I propose this Wav2vec Transducer recipe on TIMIT w/o env_corrupt.
Here is the training log:
```
epoch: 1, lr_adam: 3.00e-04, lr_wav2vec: 1.00e-04 - train loss: 9.28e-01 - valid loss: 6.01e-01, valid PER: 97.35
epoch: 2, lr_adam: 3.00e-04, lr_wav2vec: 1.00e-04 - train loss: 3.39e-01 - valid loss: 1.41e-01, valid PER: 35.27
epoch: 3, lr_adam: 3.00e-04, lr_wav2vec: 1.00e-04 - train loss: 1.73e-01 - valid loss: 9.15e-02, valid PER: 18.52
epoch: 4, lr_adam: 3.00e-04, lr_wav2vec: 1.00e-04 - train loss: 1.29e-01 - valid loss: 7.25e-02, valid PER: 16.75
epoch: 5, lr_adam: 3.00e-04, lr_wav2vec: 1.00e-04 - train loss: 1.09e-01 - valid loss: 6.30e-02, valid PER: 12.35
epoch: 6, lr_adam: 3.00e-04, lr_wav2vec: 1.00e-04 - train loss: 9.47e-02 - valid loss: 6.32e-02, valid PER: 11.24
epoch: 7, lr_adam: 3.00e-04, lr_wav2vec: 1.00e-04 - train loss: 8.53e-02 - valid loss: 6.22e-02, valid PER: 10.26
epoch: 8, lr_adam: 3.00e-04, lr_wav2vec: 1.00e-04 - train loss: 8.16e-02 - valid loss: 6.03e-02, valid PER: 10.01
epoch: 9, lr_adam: 3.00e-04, lr_wav2vec: 1.00e-04 - train loss: 7.29e-02 - valid loss: 5.86e-02, valid PER: 8.51
epoch: 10, lr_adam: 3.00e-04, lr_wav2vec: 1.00e-04 - train loss: 7.10e-02 - valid loss: 6.24e-02, valid PER: 8.43
epoch: 11, lr_adam: 3.00e-04, lr_wav2vec: 1.00e-04 - train loss: 6.65e-02 - valid loss: 6.18e-02, valid PER: 8.50
epoch: 12, lr_adam: 2.40e-04, lr_wav2vec: 9.00e-05 - train loss: 6.17e-02 - valid loss: 6.09e-02, valid PER: 8.22
epoch: 13, lr_adam: 2.40e-04, lr_wav2vec: 9.00e-05 - train loss: 5.91e-02 - valid loss: 6.49e-02, valid PER: 8.17
epoch: 14, lr_adam: 2.40e-04, lr_wav2vec: 9.00e-05 - train loss: 5.62e-02 - valid loss: 6.94e-02, valid PER: 8.29
epoch: 15, lr_adam: 1.92e-04, lr_wav2vec: 8.10e-05 - train loss: 5.48e-02 - valid loss: 6.70e-02, valid PER: 8.03
epoch: 16, lr_adam: 1.92e-04, lr_wav2vec: 8.10e-05 - train loss: 5.22e-02 - valid loss: 6.64e-02, valid PER: 7.85
epoch: 17, lr_adam: 1.92e-04, lr_wav2vec: 8.10e-05 - train loss: 4.85e-02 - valid loss: 6.80e-02, valid PER: 7.99
epoch: 18, lr_adam: 1.54e-04, lr_wav2vec: 7.29e-05 - train loss: 4.57e-02 - valid loss: 6.69e-02, valid PER: 7.97
epoch: 19, lr_adam: 1.54e-04, lr_wav2vec: 7.29e-05 - train loss: 4.37e-02 - valid loss: 6.86e-02, valid PER: 8.15
epoch: 20, lr_adam: 1.23e-04, lr_wav2vec: 6.56e-05 - train loss: 4.19e-02 - valid loss: 7.15e-02, valid PER: 8.29
Epoch loaded: 16 - test loss: 8.09e-02, test PER: 8.91
``` 
I decide to reduce the `joiner` dimension as well to fit the GPU Memory..

**Comment by aheba** [comment]
@TParcollet could you please review this PR..

File: recipes/TIMIT/ASR/seq2seq/hparams/train_with_wav2vec2.yaml
Comment (TParcollet) [inline recipes/TIMIT/ASR/seq2seq/hparams/train_with_wav2vec2.yaml:20]: To fix.

File: recipes/TIMIT/ASR/transducer/README.md
Comment (TParcollet) [inline recipes/TIMIT/ASR/transducer/README.md:18]: Please give us the model.

File: recipes/TIMIT/ASR/transducer/train_wav2vec.py
Comment (TParcollet) [inline recipes/TIMIT/ASR/transducer/train_wav2vec.py:35]: I don't see any env_corr on the YAML. Maybe simply remove all references in the train ?

File: recipes/TIMIT/ASR/transducer/train_wav2vec.py
Comment (TParcollet) [inline recipes/TIMIT/ASR/transducer/train_wav2vec.py:157]: Remove the docstring and add instead some comments on why we need to override it (dual optimizer)

File: speechbrain/decoders/transducer.py
Comment (TParcollet) [inline speechbrain/decoders/transducer.py:67]: Why this change ?

**Review by TParcollet — CHANGES_REQUESTED** [review]
Right, I checked the code, please do the requested change. @lorenlugosch could you please try the recipe, I don't have any resources right now :( So I'm only able to check the code,

File: speechbrain/decoders/transducer.py
Comment (lorenlugosch) [reply speechbrain/decoders/transducer.py:67]: The predictor is autoregressive, can't use a bidirectional net. (It's just a docstring but it's misleading)

**Comment by lorenlugosch** [comment]
@aheba shouldn't we run it with `env_corrupt` so that it's a fair comparison with the other wav2vec2 TIMIT model results?

@TParcollet I'll have a free GPU in about an hour, I'll run it then.

File: recipes/TIMIT/ASR/transducer/train_wav2vec.py
Comment (lorenlugosch) [reply recipes/TIMIT/ASR/transducer/train_wav2vec.py:35]: He took it out temporarily to save memory, see his comment above

**Comment by lorenlugosch** [comment]
correction, I may not be able to run this

**Comment by lorenlugosch** [comment]
OK, finally got TIMIT.

I'm getting this error @aheba:
```
self.proc = Wav2Vec2Processor.from_pretrained(
NameError: name 'Wav2Vec2Processor' is not defined
```

**Comment by TParcollet** [comment]
Did you install Transformer ? from hugginface ?

**Comment by lorenlugosch** [comment]
Ah OK, I ran `pip install transformers` and now it works.

Shouldn't this be in the pip requirements?

**Comment by TParcollet** [comment]
Nah, only used for wav2vec2

**Comment by lorenlugosch** [comment]
Maybe there should be a more intelligible error message if you run this recipe without HF transformers installed

**Comment by TParcollet** [comment]
There is already a try catch. I don't understand why it isn't showing ...

**Comment by lorenlugosch** [comment]
Now I get
```
numba.cuda.cudadrv.error.NvvmSupportError: libNVVM cannot be found. Do `conda install cudatoolkit`:
libnvvm.so: cannot open shared object file: No such file or directory
```

even though I ran `export NUMBAPRO_NVVM='/usr/lib/x86_64-linux-gnu/libnvvm.so'` ...

**Comment by aheba** [comment]
Thanks for reviewing, let me work on it


For the recipe with env_corrupt, my RTX8000 server is still down :/, I will recover it next week ! so if you take a look into the yaml file, I reduce the `joinder dim` for fitting the RTX6000 GPU Memory.. (at some point we need to work on the Memory issue of the RNN-T..)

File: recipes/TIMIT/ASR/seq2seq/hparams/train_with_wav2vec2.yaml
Comment (aheba) [reply recipes/TIMIT/ASR/seq2seq/hparams/train_with_wav2vec2.yaml:20]: Done

File: recipes/TIMIT/ASR/transducer/README.md
Comment (aheba) [reply recipes/TIMIT/ASR/transducer/README.md]: Done

File: recipes/TIMIT/ASR/transducer/train_wav2vec.py
Comment (aheba) [reply recipes/TIMIT/ASR/transducer/train_wav2vec.py:157]: I use the same docstring as `seq2seq/train_wav2vec.py` which explain that fit_batch override to do multiple updates

**Comment by aheba** [comment]
@TParcollet , All the reviews were addressed, LGTM

**Comment by mravanelli** [comment]
I tested this PR. It runs and the performance looks within a reasonable range of variability (a bit worse with my run, see below). LGTM if @lorenlugosch and @TParcollet are happy with it as well.

```
epoch: 1, lr_adam: 3.00e-04, lr_wav2vec: 1.00e-04 - train loss: 9.44e-01 - valid loss: 7.58e-01, valid PER: 1.00e+02
epoch: 2, lr_adam: 3.00e-04, lr_wav2vec: 1.00e-04 - train loss: 6.50e-01 - valid loss: 3.27e-01, valid PER: 73.41
epoch: 3, lr_adam: 3.00e-04, lr_wav2vec: 1.00e-04 - train loss: 2.50e-01 - valid loss: 1.10e-01, valid PER: 26.90
epoch: 4, lr_adam: 3.00e-04, lr_wav2vec: 1.00e-04 - train loss: 1.49e-01 - valid loss: 8.37e-02, valid PER: 19.21
epoch: 5, lr_adam: 3.00e-04, lr_wav2vec: 1.00e-04 - train loss: 1.21e-01 - valid loss: 7.29e-02, valid PER: 15.36
epoch: 6, lr_adam: 3.00e-04, lr_wav2vec: 1.00e-04 - train loss: 1.03e-01 - valid loss: 6.49e-02, valid PER: 10.91
epoch: 7, lr_adam: 3.00e-04, lr_wav2vec: 1.00e-04 - train loss: 9.65e-02 - valid loss: 6.19e-02, valid PER: 11.03
epoch: 8, lr_adam: 2.40e-04, lr_wav2vec: 9.00e-05 - train loss: 8.59e-02 - valid loss: 5.98e-02, valid PER: 10.61
epoch: 9, lr_adam: 2.40e-04, lr_wav2vec: 9.00e-05 - train loss: 7.93e-02 - valid loss: 5.94e-02, valid PER: 9.19
epoch: 10, lr_adam: 2.40e-04, lr_wav2vec: 9.00e-05 - train loss: 7.53e-02 - valid loss: 5.86e-02, valid PER: 9.03
epoch: 11, lr_adam: 2.40e-04, lr_wav2vec: 9.00e-05 - train loss: 7.05e-02 - valid loss: 6.04e-02, valid PER: 8.79
epoch: 12, lr_adam: 2.40e-04, lr_wav2vec: 9.00e-05 - train loss: 6.64e-02 - valid loss: 6.14e-02, valid PER: 8.34
epoch: 13, lr_adam: 2.40e-04, lr_wav2vec: 9.00e-05 - train loss: 6.44e-02 - valid loss: 6.18e-02, valid PER: 8.30
epoch: 14, lr_adam: 2.40e-04, lr_wav2vec: 9.00e-05 - train loss: 6.29e-02 - valid loss: 6.44e-02, valid PER: 8.15
epoch: 15, lr_adam: 2.40e-04, lr_wav2vec: 9.00e-05 - train loss: 5.91e-02 - valid loss: 6.63e-02, valid PER: 8.38
epoch: 16, lr_adam: 1.92e-04, lr_wav2vec: 8.10e-05 - train loss: 5.66e-02 - valid loss: 6.34e-02, valid PER: 8.22
epoch: 17, lr_adam: 1.92e-04, lr_wav2vec: 8.10e-05 - train loss: 5.11e-02 - valid loss: 6.58e-02, valid PER: 7.76
epoch: 18, lr_adam: 1.92e-04, lr_wav2vec: 8.10e-05 - train loss: 5.04e-02 - valid loss: 6.71e-02, valid PER: 8.00
epoch: 19, lr_adam: 1.54e-04, lr_wav2vec: 7.29e-05 - train loss: 4.79e-02 - valid loss: 6.62e-02, valid PER: 7.78
epoch: 20, lr_adam: 1.54e-04, lr_wav2vec: 7.29e-05 - train loss: 4.68e-02 - valid loss: 6.78e-02, valid PER: 7.91
Epoch loaded: 17 - test loss: 8.36e-02, test PER: 9.41
```

**Comment by aheba** [comment]
yeah, @mravanelli , I have a small gpu.. so I constraint the joint_dim to be 39dim.. I should go down with a more larger dim.

@TParcollet , @lorenlugosch  If you are ok, LTGM

**Comment by TParcollet** [comment]
@aheba does not look like reviews were adressed to me ? I still see the issues in the changed files.

**Comment by aheba** [comment]
could you please @TParcollet share the issues ?

**Comment by TParcollet** [comment]
See above, things have been marked as resolved but I still see /scratch/TIMIT for example.

**Comment by aheba** [comment]
@TParcollet , fixed.. it was the seq2seq recipes..

Do you see the CI errors on CommonVoice recipe.. @pplantinga  do you have an idea why it generate these errors ?

**Comment by TParcollet** [comment]
@mravanelli is on it (for the errors)

**Comment by mravanelli** [comment]
I saw that environmental corruption has been removed from train_wav2vect.py. I think we should keep it because people might wanna use it even with wav2vect, right? This is the same of other recipes...

**Comment by mravanelli** [comment]
The same comment actually applies to all the wav2vect recipes where env_corruption has been removed...

**Comment by TParcollet** [comment]
Nah, I don't think it is relevant. The training is already hugely VRAM-consuming and the env corruption does not add anything to the results from what I saw.

**Comment by mravanelli** [comment]
ok, can we merge then?

**Comment by TParcollet** [comment]
Yes.

**Pull request merged**
